File: apps/robotlab_x/src/robotlab_x/models/message.py
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict
from typing import List, Any, Optional, Literal, Dict
import time
import json
import yaml
from database.types import DateTimeStr

class Message(BaseModel):
    id: Optional[str] = Field(None, description="Transcription Request ID - can be created by client or will be self generated", json_schema_extra={"example":"123e4567-e89b-12d3-a456-426614174000"})
    name: Optional[str] = Field(None, description="the unique service name / topic — None would be simple broadcast", json_schema_extra={"example":"ollama@rbx.com"})
    type: Optional[str] = Field(None, description="service")
    method: str = Field(None, description="the method to be invoked on the service", json_schema_extra={"example":"visit-1234"})
    data: Optional[dict] = Field(None, description="the data to be sent to the service", json_schema_extra={"type":"object","additionalProperties":True,"example":{"key":"value"}})
    reply_to: Optional[str] = Field(None, description="the message ID to reply to for request-response patterns", json_schema_extra={"example":"123e4567-e89b-12d3-a456-426614174000"})

# managed
    @classmethod
    def from_yaml(cls, yaml_str: str):
        """Load Message from a YAML string."""
        data = yaml.safe_load(yaml_str)
        return cls(**data)

    @classmethod
    def from_yaml_file(cls, file_path: str):
        """Load Message from a YAML file."""
        with open(file_path, "r") as file:
            return cls.from_yaml(file.read())

    @classmethod
    def from_json(cls, json_str: str):
        """Load Message from a JSON string."""
        data = json.loads(json_str)
        return cls(**data)

    @classmethod
    def from_json_file(cls, file_path: str):
        """Load Message from a JSON file."""
        with open(file_path, "r") as file:
            return cls.from_json(file.read())

File: apps/robotlab_x_ui/src/models/Message.ts
// managed
// Message interface for use across the project
export interface Message {
  id?: string;
  name?: string;
  type?: string;
  method: string;
  data?: any;
  reply_to?: string;
}

export function createEmptyMessage(): Message {
  return {
    id: undefined,
    name: undefined,
    type: undefined,
    method: "",
    data: undefined,
    reply_to: undefined,
  };
}

// Helper function to validate Message object
export function isMessage(obj: any): obj is Message {
  return obj && typeof obj === 'object';
}

// Helper function to clone Message object
export function cloneMessage(item: Message): Message {
  return {
    ...item,
  };
}

// JSON Schema for Fastify route validation and OpenAPI documentation
export const MessageSchema = {
  "type": "object",
  "properties": {
    "id": {
      "type": "string",
      "description": "Transcription Request ID - can be created by client or will be self generated",
      "example": "123e4567-e89b-12d3-a456-426614174000"
    },
    "name": {
      "type": "string",
      "description": "the unique service name / topic \u2014 None would be simple broadcast",
      "example": "ollama@rbx.com"
    },
    "type": {
      "type": "string",
      "description": "service"
    },
    "method": {
      "type": "string",
      "description": "the method to be invoked on the service",
      "example": "visit-1234"
    },
    "data": {
      "description": "the data to be sent to the service",
      "example": {
        "key": "value"
      },
      "type": "object",
      "additionalProperties": true
    },
    "reply_to": {
      "type": "string",
      "description": "the message ID to reply to for request-response patterns",
      "example": "123e4567-e89b-12d3-a456-426614174000"
    }
  },
  "required": [
    "method"
  ]
}
